### PR TITLE
Enable perlcritic checks in CI

### PR DESCRIPTION
**Pull request by b10n1k**
Run perlcritic against all the perl git files.
    
    https://progress.opensuse.org/issues/138416

### CODE REVIEW

**Comment by perlpunk** [comment]
I think the git commit message checker can be replaced with what we have in openQA: https://github.com/os-autoinst/openQA/blob/master/.github/workflows/commit-message-checker.yml
I think it also has an exception for git subrepo commit messages.

**Comment by b10n1k** [comment]
i guess the errors are not by my PR??!

**Comment by perlpunk** [comment]
I think the isotovideo tests have been broken for a long time. I don't know anything about them, I think they were even there before I joined. Maybe someone else can say something about it, but I would say they can be ignored for this PR.
But the git commit message check can be fixed by what I said.

**Comment by b10n1k** [comment]
> I think the git commit message checker can be replaced with what we have in openQA: https://github.com/os-autoinst/openQA/blob/master/.github/workflows/commit-message-checker.yml I think it also has an exception for git subrepo commit messages.

i will have to take a look on that

File: .github/workflows/check_critic.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    runs-on: ubuntu-latest
+    container:
+      image: registry.opensuse.org/devel/openqa/containers/os-autoinst_dev
+    steps:
+      - uses: actions/checkout@v4
+      - name: Static analysis
+        run: |
+          git config --global --add safe.directory '*'
+          zypper -n install perl-Perl-Critic
+          perlcritic $(git ls-files -- '*.p[ml]')
```
Comment (b10n1k) [inline .github/workflows/check_critic.yml:16]: i wonder whether i have to use https://github.com/os-autoinst/os-autoinst-common/blob/master/tools/perlcritic somehow here

Comment (perlpunk) [reply .github/workflows/check_critic.yml:16]: Maybe @josegomezr or @okurz can answer that

Comment (josegomezr) [reply .github/workflows/check_critic.yml:16]: `perlcritic` kinda "knows" what's a perl file, you can pass just the directories where it needs to be invoked.

In fact, the check is in the files you're pulling from `os-autoinst-common`. Use that as much as possible.

```yaml
# [...]
container:
      image: perldocker/perl-tester # this is now provided by os_autoinst dev container iirc
    steps:
      - uses: actions/checkout@v4
      - run: ./tools/perlcritic --quiet 
# [...]
```

**Comment by Martchus** [comment]
Yes, the isotovideo checks were failing before as @perlpunk mentioned. And for the commit message check see her comment as well.

**Review by Martchus — APPROVED** [review]

**Comment by b10n1k** [comment]
please do not merge yet. I am working on isotovideo failures

**Comment by josegomezr** [comment]
@b10n1k "github action example / test (pull_request)" and "check all test modules / test (pull_request)" always fail, don't ask me why.

#29 tackles a bit that so they are more native and run in a sensible manner in case you need another example around ;)

**Comment by perlpunk** [comment]
Yeah, you don't need to fix the isotovideo tests in this PR. sorry if that wasn't clear

**Review by Martchus — APPROVED** [review]
Nice that you've also been able to fix the isotovideo checks.

File: .github/workflows/isotovideo-check-all-test-modules.yml [hunk 2]
```diff
@@ -13,4 +13,9 @@ jobs:
     steps:
       - uses: actions/checkout@v2
       - name: Run isotovideo against test code, fail if any test module failed
-        run: podman run --rm -it -v .:/tests:Z --entrypoint '' registry.opensuse.org/devel/openqa/containers/isotovideo:qemu-x86-jq /bin/sh -c 'isotovideo qemu_no_kvm=1 casedir=/tests && jq .result testresults/result-*.json | grep -v ok && echo "Test modules failed" && exit 1'
+        env:
+          image: registry.opensuse.org/devel/openqa/containers/isotovideo:qemu-x86-jq
+          isotovideo: isotovideo qemu_no_kvm=1 casedir=/tests
+          jq_filter: jq .result testresults/result-*.json
+          err_msg: Test modules failed
+        run: docker run --rm -it -v .:/tests:Z --entrypoint '' $image /bin/sh -c '$isotovideo && $jq_filter | grep ok || (echo "$err_msg" && exit 1)'
```
Comment (baierjan) [inline .github/workflows/isotovideo-check-all-test-modules.yml]: How is this change related to perlcritic checks?

Is the switch from podman to docker intended? I thought podman is the preferred way.

Also, you probably do not want the exit in subshell and rather use `grep ok || { echo "$err_msg" && exit 1; }`

File: .github/workflows/yaml-checks.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  yamllint:
+    runs-on: ubuntu-latest
+    container:
+      image: registry.opensuse.org/devel/openqa/containers/os-autoinst_dev
+    steps:
+      - uses: actions/checkout@v4
+      - name: Validate yamls
+        run: |
+          git config --global --add safe.directory '*'
+          yamllint --strict $(git ls-files "*.yml" "*.yaml" 2> /dev/null || find . -name '*.y*ml')
```
Comment (baierjan) [inline .github/workflows/yaml-checks.yml:15]: `*.y*ml` will match a lot of different things, maybe you want `find . -name '*.yaml' -o -name '*.yml'` to have closer match to the list used in git? Also `2> /dev/null` might be good addition for the find as well.

File: .github/workflows/isotovideo-action.yml [hunk 4]
```diff
@@ -9,11 +9,9 @@ jobs:
       image: "registry.opensuse.org/devel/openqa/containers/isotovideo:qemu-x86-jq"
     steps:
       - uses: actions/checkout@v2
```
Comment (baierjan) [inline .github/workflows/isotovideo-action.yml:11]: Can you also change the action to `actions/checkout@v4`?

File: .github/workflows/isotovideo-action.yml [hunk 5]
```diff
@@ -9,11 +9,9 @@ jobs:
...
-      - name: install jq
-        run: zypper -n in jq
 
       - name: Run isotovideo against test code
-        run: isotovideo qemu_no_kvm=1 casedir=.
+        run: isotovideo qemu_no_kvm=1 casedir=$(pwd)
 
       - name: fail if any test module failed
-        run: jq .result testresults/result-*.json | grep -v ok && echo "Test modules failed" && exit 1
+        run: jq .result testresults/result-*.json | grep ok || (echo "Test modules failed" && exit 1)
```
Comment (baierjan) [inline .github/workflows/isotovideo-action.yml:17]: As mentioned elsewhere, you probably do not want subshell here, the exit will only exit the subshell.
`grep ok || { echo "Test modules failed" && exit 1; }`

File: .github/workflows/isotovideo.yml [hunk 6]
```diff
@@ -8,4 +8,4 @@ jobs:
     steps:
       - uses: actions/checkout@v2
       - name: Run isotovideo against test code in happy-path scenario
-        run: podman run --rm -it -v .:/tests:Z registry.opensuse.org/devel/openqa/containers/isotovideo:qemu-x86 qemu_no_kvm=1 casedir=/tests
+        run: docker run --rm -v .:/tests:Z registry.opensuse.org/devel/openqa/containers/isotovideo:qemu-x86 qemu_no_kvm=1 casedir=/tests
```
Comment (baierjan) [inline .github/workflows/isotovideo.yml:11]: I understand the removal of `-i` and `-t`, why the podman -> docker change?

File: .github/workflows/isotovideo-check-all-test-modules.yml [hunk 7]
```diff
@@ -13,9 +13,9 @@ jobs:
...
-          isotovideo: isotovideo qemu_no_kvm=1 casedir=/tests
-          jq_filter: jq .result testresults/result-*.json
-          err_msg: Test modules failed
-        run: docker run --rm -it -v .:/tests:Z --entrypoint '' $image /bin/sh -c '$isotovideo && $jq_filter | grep ok || (echo "$err_msg" && exit 1)'
+        run: |
+          docker run --rm -v .:/tests:Z --entrypoint '' \
+          registry.opensuse.org/devel/openqa/containers/isotovideo:qemu-x86-jq /bin/sh \
+          -c 'isotovideo qemu_no_kvm=1 casedir=/tests \
+          && jq .result testresults/result-*.json | grep ok \
+          || (echo "Test modules failed" && exit 1)'
```
Comment (baierjan) [inline .github/workflows/isotovideo-check-all-test-modules.yml:21]: Again the subshell.

**Review by baierjan — CHANGES_REQUESTED** [review]

File: .github/workflows/isotovideo-action.yml (hunk 4 above)
Comment (baierjan) [reply .github/workflows/isotovideo-action.yml:11]: That change should get rid of the deprecation warning seen in https://github.com/os-autoinst/os-autoinst-distri-example/actions/runs/7831480244

File: .github/workflows/isotovideo-action.yml (hunk 5 above)
Comment (perlpunk) [inline .github/workflows/isotovideo-action.yml:17]: Previously the grep was looking for *any* line not matching `ok`, so any failed test would make the workflow fail.
Now you are only ensuring that there is at least one line matching `ok`. So there could be failed modules and it would still succeed.
I don't know how many testmodules this workflow runs, but I think the logic should be kept as before, or look at @josegomezr ' PR.

Comment (baierjan) [reply .github/workflows/isotovideo-action.yml:17]: Ah, thanks for commenting that. I also noticed that but forget to note down that `grep -v ok` is not the exact opposite of `grep ok` (I got sidetracked by the subshell 😄 ). I also vote for the previous logic; I believe the motivation was to avoid the non-zero exit code provided by grep on "success". We might want to use `grep -v ok && { echo "Test modules failed" && exit 1; } || true`

**Review by perlpunk — COMMENTED** [review]
There are already a lot of comments in this PR, so I suggest to keep the isotovideo checks out of here. It's unrelated and should be in its own PR.

**Review by okurz — CHANGES_REQUESTED** [review]
Please split this PR into separate ones based on the commits you created. Consider https://progress.opensuse.org/projects/qa/wiki/Tools#Collaboration-best-practices

After that I would like to review the individual parts in more details

**Comment by b10n1k** [comment]
this is outdated. other PRs provided and merged
